flow-osgi/src/main/java/com/vaadin/flow/osgi/support/AppConfigFactoryTracker.java

```diff
@@ -0,0 +1,377 @@
+        }
+
+        @Override
+        public <T> Collection<T> lookupAll(Class<T> serviceClass) {
+            return OSGiVaadinInitialization.lookupAll(getWebAppBundle(),
+                    serviceClass);
+        }
+
+        private Bundle getWebAppBundle() {
+            return webAppBundle;
+        }
+
+    }
+
+    AppConfigFactoryTracker(Bundle webAppBundle, VaadinServletContext context,
+            ServletContainerInitializerClasses initializerClasses)
+            throws IllegalContextState {
+        super(webAppBundle.getBundleContext(),
+                ApplicationConfigurationFactory.class, null);
+        this.webAppBundle = webAppBundle;
+        servletContext = context;
+        this.initializerClasses = initializerClasses;
+        webAppBundle.getBundleContext().addBundleListener(this);
+    }
+
+    @Override
+    public ApplicationConfigurationFactory addingService(
+            ServiceReference<ApplicationConfigurationFactory> reference) {
+        ApplicationConfigurationFactory factory = super.addingService(
+                reference);
+        AppConfigFactoryTracker tracker = servletContext
+                .getAttribute(AppConfigFactoryTracker.class);
+        if (tracker != null) {
+            stop();
+            servletContext.removeAttribute(AppConfigFactoryTracker.class);
+        }
+        initializeLookup();
+        return factory;
+    }
+
+    @Override
+    public void bundleChanged(BundleEvent event) {
+        if ((event.getType() & BundleEvent.STOPPED) != 0) {
+            stop();
+        }
+    }
+
+    private void stop() {
+        close();
+        context.removeBundleListener(this);
+    }
+
+    private void initializeLookup() {
+        /*
+         * There are two cases:
+         * 
+         * - the servlet context is initialized before all of the Servlets or
+         * 
+         * - after at least one of the servlets with the same context.
+         * 
+         * SevletContext::getClassLoader doesn't return the container the web
+         * app classloader for the servlet context instance here (in the
+         * "contextInitialized" method). It returns the web app classloader for
+         * the servlet context which is passed to the Servlet::init method: in
+         * fact two servlet context instances are different even though they
+         * share the same attributes and values. So to be able to identify the
+         * servlet context (which is coming from Servlet) and the
+         * "servletContext" instance here the Lookupx instance is used which is
+         * set as an attribute and if the context "is the same" the attributes
+         * should be the same.
+         */
+
+        // at this point at least one of the Servlet::init method should be
+        // called and the classloader should be available
+
+        // ensure the lookup is set into the context
+        Lookup created[] = new Lookup[1];
```

Comment:
Do you want to fix this from C syntax to Java syntax?

Reply:
Do you mean `Lookup[]`  vs `created[]`  ?
I already don't remember which is Java which is C.

This is important in a field declaration.
Here it's just a tiny impl detail to get value from a lambda.
It could have been a `List`  instead. But `List` just needs more memory than array.

Reply:
C marks it on a declarator and Java on the type
`Lookup[] created` is Java standard
`Lookup created[]` is from C and I didn't even know it was supported in Java =)

Reply:
It's supported , but recommendation is not use it.

Done.